[Modifies Set] Implemented modifies set inference.

kkmc opened the pull request:
Implemented modifies set analysis passes that can be turned on with the -M option.

The first pass infers the modifies sets for each procedure and stores them in a map from procedure id to sets.

The second pass will combine (1) the original modifies set, (2) the inferred modifies set, and (3) the modifies sets of the callees.

Includes 4 test cases.

NOTE: This does not work with instance procedure calls.

NOTE 2: Passed config into the compile function in UclidMain. As a result, I removed the file argument in the compile function, which changes how all test cases need to be instantiated (you need to specify a Config object instead of a Sequence of File objects). I create a file called Utils.scala in the test directory that contains helper functions for creating these Config files with different options to the solver.

polgreen commented:
I'll look over this properly tomorrow, but first comments: 
- I think you're only testing the parsing and not the functionality here (I think the tests are only run in parserspec?). Is that right? Why?
- What happens when you do this with instance procedure calls? Does it throw an error message, or produce incorrect results?

kkmc commented:
@polgreen 

> I think you're only testing the parsing and not the functionality here (I think the tests are only run in parserspec?). Is that right? Why?

Added another test to check that it won't add more modified variables than necessary.

> What happens when you do this with instance procedure calls? Does it throw an error message, or produce incorrect results?

Just added a line to throw an unimplemented error for that case.

polgreen commented:
I think there's an issue with local variables if their ID clashes with a global variable, e.g., the correct modifies set for bar here is just `x` because the assignment to `y` is to the local var, but mod set inference is inferring that the set is `x` and `y`
- If you run this file without mod set inference and with the modifies sets that I've written in, verification passes. 
- If you comment out the modifies sets and run with `-M` verification fails

~~~
module main {

    var x: integer;
    var y: integer;

    procedure [noinline] bar()
        ensures x == old(x) + 1;
        modifies x;
    {
      var y: integer;
       x = x+1; 
       y = 1; // this is a local var assignment
    }

    // Check that calling bar does not change y
    procedure foo()
        ensures y == old(y);
        ensures x == old(x) + 1;
        modifies x;
    {
        call bar();
    }

    control {
        v = verify(bar);
        v = verify(foo);
        check;
        print_results;
    }
}
~~~

polgreen reviewed (changes requested):
OK, the way it is implemented looks fine to me

But there are some issues to fix (can you add tests for them when you fix them please?): 
- module variables that have the same name as local variables changes the result (test = the one i posted in the comment above)
- test/test-inliner.ucl throws an error. I think this is something to do with when local variables are used as the return values from procedure calls. e.g.,
~~~
    var s1 : bit_t;
    var c1 : bit_t;
    call (s1, c1) = half_adder1(a, b);
~~~
- test/test-tuple-record-1.ucl: verification passes without mod set inference and fails with. It's inferring that the global variable s is being modified, when actually it's the return variable being modified, which shares the same name as the global.

kkmc commented on src/test/scala/VerifierSpec.scala:
```diff
@@ -45,15 +45,16 @@ import java.io.File
 
 object VerifierSpec {
-  def expectedFails(filename: String, nFail : Int) : String = {
+  def expectedFails(filename: String, nFail : Int, config: Option[UclidMain.Config]=None) : String = {
```
@polgreen  Is there another way to provide a different config to the tests? I'm trying to use the -M option here.

polgreen replied:
No, I don't think so, but it would be nice if we were consistent across the specs. it's fine how it is, I might change it so the other specs match

kkmc commented on src/test/scala/VerifierSpec.scala:
```diff
@@ -45,15 +45,16 @@ import java.io.File
-    val mainModule = UclidMain.instantiate(config, modules, l.Identifier("main"), false)
+    val compileConfig = if (config.isDefined) config.get else ConfigCons.createConfig(filename)
+    val modules = UclidMain.compile(compileConfig, lang.Identifier("main"), true)
+    val instantiateConfig = UclidMain.Config() 
```
I wanted to allow different configurations to be passed. so there's a compileConfig that depends on whether we pass `some` config to the function. This was done for (-M) options.

kkmc commented:
@polgreen 

Thanks for the review! I had forgotten that UCLID supports shadowing. Fixed by keeping track of the local variable set at each scope and not including the changes to those locals.

Other details regarding your review:
(1) Added your test case
(2) test/test-inliner.ucl and test/test-tuple-record-1.ucl are fixed by the shadow fix
